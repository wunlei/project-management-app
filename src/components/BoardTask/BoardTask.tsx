import { useState } from 'react';
import {
  Avatar,
  Chip,
  IconButton,
  Menu,
  MenuItem,
  Stack,
  Typography,
} from '@mui/material';
import { ReactComponent as ClipIcon } from 'assets/icons/paperclip.svg';
import { ReactComponent as MenuIcon } from 'assets/icons/menu.svg';
import { BoardTaskProps } from './BoardTask.types';

function BoardTask({ title, description, isDone, user }: BoardTaskProps) {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <Stack
      spacing={0}
      padding={1.5}
      borderRadius={1}
      bgcolor="#ffffff"
      sx={{
        '&:hover': {
          cursor: 'pointer',
          boxShadow: 2,
        },
      }}
      width="250px"
      margin="5px"
    >
      <Stack direction="row" justifyContent="space-between" alignItems="center">
        {/* fix colors contrast */}
        <Chip
          label={isDone ? 'Done' : 'Not Done'}
          color={isDone ? 'success' : 'warning'}
          size="small"
          sx={{
            fontSize: '0.875rem',
          }}
        />
        <div>
          <IconButton
            id="menu-button"
            aria-controls={open ? 'card-menu' : undefined}
            aria-haspopup="true"
            aria-expanded={open ? 'true' : undefined}
            onClick={handleClick}
          >
            <MenuIcon />
          </IconButton>
          <Menu
            id="card-menu"
            anchorEl={anchorEl}
            open={open}
            onClose={handleClose}
            MenuListProps={{
              'aria-labelledby': 'menu-button',
            }}
          >
            <MenuItem onClick={handleClose}>Open</MenuItem>
            <MenuItem onClick={handleClose}>Edit</MenuItem>
            <MenuItem onClick={handleClose}>Delete</MenuItem>
          </Menu>
        </div>
      </Stack>
      <Typography variant="h6" fontWeight="bold">
        {title}
      </Typography>
      <Stack direction="row" alignItems="center" justifyContent="space-between">
        <ClipIcon width="20px" height="20px" />
        <Avatar sx={{ bgcolor: 'orange', width: '30px', height: '30px' }}>
          {user}
        </Avatar>
      </Stack>
    </Stack>
  );
}

export default BoardTask;
