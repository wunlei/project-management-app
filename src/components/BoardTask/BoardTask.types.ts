export interface BoardTaskProps {
  title: string;
  description?: string;
  // isDone: boolean;
  user: string;
}
