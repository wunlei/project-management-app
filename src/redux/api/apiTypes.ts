export type UserFromServer = { id: string; name: string; login: string };
export type UserFromClient = { name: string; login: string; password: string };

export type FileInfo = {
  filename: string;
  fileSize: number;
};

export type TaskFromServer = {
  id: string;
  title: string;
  order: number;
  done: boolean;
  description: string;
  userId: string;
  files: Array<FileInfo>;
};
export type TaskFromServerExpanded = TaskFromServer & {
  boardId: string;
  columnId: string;
};
export type TaskFromClient = {
  title: string;
  order?: number;
  done?: boolean;
  description: string;
  userId?: string;
};
export type TaskFromClientExpanded = TaskFromClient & {
  boardId: string;
  columnId: string;
};

export type ColumnFromServer = {
  id: string;
  title: string;
  order: number;
};
export type ColumnFromServerExpended = ColumnFromServer & {
  tasks: Array<TaskFromServer>;
};
export type ColumnFromClient = { title: string; order: number };

export type BoardFromServer = { id: string; title: string };
export type BoardFromServerExpanded = BoardFromServer & {
  columns: Array<ColumnFromServerExpended>;
};

// Sign -----

export type SignInResult = { token: string; id: string };
export type SignInArg = { body: { login: string; password: string } };

export type SignUpResult = UserFromServer;
export type SignUpArg = { body: UserFromClient };

// Users -----

export type GetAllUsersResult = Array<UserFromServer>;
export type GetAllUsersArg = void;

export type GetUserResult = UserFromServer;
export type GetUserArg = { userId?: string };

export type DeleteUserResult = void;
export type DeleteUserArg = { userId?: string };

export type UpdateUserResult = UserFromServer;
export type UpdateUserArg = { userId?: string; body: UserFromClient };

// Boards -----

export type GetAllBoardsResult = Array<BoardFromServer>;
export type GetAllBoardsArg = void;

export type GetBoardResult = BoardFromServerExpanded;
export type GetBoardArg = { boardId: string };

export type CreateBoardResult = BoardFromServer;
export type CreateBoardArg = { body: { title: string } };

export type DeleteBoardResult = void;
export type DeleteBoardArg = { boardId: string };

export type UpdateBoardResult = BoardFromServer;
export type UpdateBoardArg = { boardId: string; body: { title: string } };

// Columns -----

export type GetAllColumnsResult = Array<ColumnFromServer>;
export type GetAllColumnsArg = { boardId: string };

export type GetColumnResult = ColumnFromServerExpended;
export type GetColumnArg = { boardId: string; columnId: string };

export type CreateColumnResult = ColumnFromServer;
export type CreateColumnArg = {
  boardId: string;
  body: ColumnFromClient;
};

export type DeleteColumnResult = void;
export type DeleteColumnArg = { boardId: string; columnId: string };

export type UpdateColumnResult = ColumnFromServer;
export type UpdateColumnArg = {
  boardId: string;
  columnId: string;
  body: ColumnFromClient;
};

// Tasks -----

export type GetAllTasksResult = Array<TaskFromServerExpanded>;
export type GetAllTasksArg = {
  boardId: string;
  columnId: string;
};

export type GetTaskResult = TaskFromServerExpanded;
export type GetTaskArg = {
  boardId: string;
  columnId: string;
  taskId: string;
};

export type CreateTaskResult = TaskFromServerExpanded;
export type CreateTaskArg = {
  boardId: string;
  columnId: string;
  body: TaskFromClient;
};

export type DeleteTaskResult = void;
export type DeleteTaskArg = {
  boardId: string;
  columnId: string;
  taskId: string;
};

export type UpdateTaskResult = TaskFromServerExpanded;
export type UpdateTaskArg = {
  boardId: string;
  columnId: string;
  taskId: string;
  body: TaskFromClientExpanded;
};

// File -----

export type UploadFileResult = void;
export type UploadFileArg = {
  taskId: string;
  file: File;
};
