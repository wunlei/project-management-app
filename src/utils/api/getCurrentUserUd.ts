import UserDeletedLocalStorageError from 'utils/errors/UserDeletedLocalStorageError';

export default function getCurrentUserId() {
  const userId = localStorage.getItem('userId');
  if (!userId) throw new UserDeletedLocalStorageError();

  return userId;
}
