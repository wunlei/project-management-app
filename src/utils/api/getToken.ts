export default function getToken() {
  const token = localStorage.getItem('token');
  if (!token) throw new Error('Token is missing in local storage');

  return token;
}
