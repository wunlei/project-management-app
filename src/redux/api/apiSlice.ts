import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import {
  BaseQueryFn,
  FetchArgs,
  FetchBaseQueryError,
} from '@reduxjs/toolkit/query';
import getToken from 'utils/api/getToken';
import UserDeletedLocalStorageError from 'utils/errors/UserDeletedLocalStorageError';
import { setUserId, setToken } from '../global/globalSlice';

export const baseUrl = 'http://localhost:4000';

const baseQuery = fetchBaseQuery({
  baseUrl,
  prepareHeaders: (headers, { endpoint }) => {
    const endpointsWithoutToken = ['signIn', 'signUp'];

    if (endpointsWithoutToken.includes(endpoint)) {
      return headers;
    }

    headers.set('authorization', `Bearer ${getToken()}`);
    return headers;
  },
});

const baseQueryWithErrorHandlers: BaseQueryFn<
  string | FetchArgs,
  unknown,
  FetchBaseQueryError
> = async (args, api, extraOptions) => {
  try {
    const result = await baseQuery(args, api, extraOptions);

    if (result.error && result.error.status === 401) {
      api.dispatch(setUserId(null));
      api.dispatch(setToken(null));
    }

    return result;
  } catch (error) {
    if (error instanceof UserDeletedLocalStorageError) {
      api.dispatch(setUserId(null));
      api.dispatch(setToken(null));
      window.location.reload();
      throw new UserDeletedLocalStorageError();
    }
    throw new Error('Arbitrary error');
  }
};

export const apiSlice = createApi({
  reducerPath: 'api',
  baseQuery: baseQueryWithErrorHandlers,
  tagTypes: [
    'USER',
    'USERS',
    'BOARD',
    'BOARDS',
    'COLUMN',
    'COLUMNS',
    'TASK',
    'TASKS',
  ],
  endpoints: () => ({}),
});
