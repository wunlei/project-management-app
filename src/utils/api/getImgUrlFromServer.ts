import { baseUrl } from 'redux/api/apiSlice';
import getToken from './getToken';

export default function getImgUrlFromServer({
  taskId,
  filename,
}: {
  taskId: string;
  filename: string;
}) {
  return new Promise<string>((resolve, reject) => {
    fetch(`${baseUrl}/file/${taskId}/${filename}`, {
      headers: {
        Authorization: `Bearer ${getToken()}`,
      },
    })
      .then((response) => {
        if (!response.body) throw new Error('No file');

        const reader = response.body.getReader();

        return new ReadableStream({
          start(controller) {
            return pump();

            function pump(): unknown {
              return reader.read().then(({ done, value }) => {
                if (done) {
                  controller.close();
                  return;
                }

                controller.enqueue(value);
                return pump();
              });
            }
          },
        });
      })
      .then((stream) => new Response(stream))
      .then((response) => response.blob())
      .then((blob) => resolve(URL.createObjectURL(blob)))
      .catch((err) => reject(err));
  });
}
