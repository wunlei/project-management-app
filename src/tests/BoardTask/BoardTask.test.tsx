import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import BoardTask from 'components/BoardTask/BoardTask';
import { TaskFromServerExpanded } from 'redux/api/apiTypes';

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    };
  },
}));

it('should render with provided data', () => {
  const data = {
    user: 'Name',
    task: {},
    handleTaskEditModalOpen: () => {},
    handleTaskDeleteConfirmOpen: () => {},
  };
  const card = renderer
    .create(
      <BrowserRouter>
        <BoardTask
          user={data.user}
          task={data.task as TaskFromServerExpanded}
          handleTaskEditModalOpen={data.handleTaskEditModalOpen}
          handleTaskDeleteConfirmOpen={data.handleTaskDeleteConfirmOpen}
        />
      </BrowserRouter>
    )
    .toJSON();
  expect(card).toMatchSnapshot();
});
