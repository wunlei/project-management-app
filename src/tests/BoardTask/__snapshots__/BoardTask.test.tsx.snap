// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with provided data 1`] = `
<div
  className="css-rbqpdu-MuiStack-root"
  onClick={[Function]}
>
  <div
    className="css-1hvyxw4-MuiStack-root"
  >
    <h6
      className="MuiTypography-root MuiTypography-h6 css-mzbhic-MuiTypography-root"
    />
    <button
      aria-label="Delete"
      aria-labelledby={null}
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
      data-mui-internal-clone-element={true}
      disabled={false}
      id="menu-button"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        style={
          Object {
            "height": "18px",
            "width": "18px",
          }
        }
      >
        trash.svg
      </svg>
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    className="css-1gb3guw-MuiStack-root"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1vyukf0-MuiAvatar-root"
    >
      N
    </div>
  </div>
</div>
`;
